compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -130,7 +125,8 @@ pub fn intern_const_alloc_recursive<
 
     // Initialize recursive interning.
     let base_alloc_id = ret.ptr().provenance.unwrap().alloc_id();
-    let mut todo = vec![(base_alloc_id, base_mutability)];
+    let mut todo: Vec<_> =
+        intern_shallow(ecx, base_alloc_id, base_mutability).unwrap().map(|prov| prov).collect();
     // We need to distinguish "has just been interned" from "was already in `tcx`",
     // so we track this in a separate set.
     let mut just_interned = FxHashSet::default();
```

Comment:
You need to add base_alloc_id here (i.e., you forgot to duplicate part of the code)

Reply:
I'll try building a test for it

Reply:
I don't think that's possible. It might be that your code is fine. But it is certainly different from the current logic in master and there's no comment explaining why the first allocation should not be in that set.

Reply:
Resolved by just nuking the `just_interned` set entirely as it was already dead code

Reply:
I think you nuked too much, `just_interned` was part of a [key soundness check](https://github.com/rust-lang/rust/pull/120302#discussion_r1467494847). It might be a redundant check, but I still consider it important. That check is the entire reason why I rewrote the interner.

Reply:
iI mmemeant that it was dead code statically. Not dynamically. We inserted an id in the same loop as we checked it

Reply:
Of course that was only due to my refactorings. But these refactorings were just "put all alloc ids into the todo list, filter in iteration" instead of the previous "filter before inserting into todo list"

I do not believe I made a functional change

Reply:
Filtering before inserting makes a big difference, it means we do the `try_get_global_alloc` check *after* the checks in the closure.